File: Models/Period.cs
using System;
using JetBrains.Annotations;

namespace Models
{
    /// <summary>
    /// Period when composer lived and worked, like Late Baroque or Romanticism. 
    /// </summary>
    [PublicAPI]
    public record Period
    {
        /// <summary>
        /// Id in the database.
        /// </summary>
        public int Id { get; init; }

        /// <summary>
        /// Name of the period.
        /// </summary>
        /// <example>Late Baroque</example>
        public string Name { get; init; } = "";
        
        /// <summary>
        /// Year when period roughly started.
        /// </summary>
        public int YearStart { get; init; }
        
        /// <summary>
        /// Year when period roughly ended. Can be null for the lasted period which is still ongoing.
        /// </summary>
        public int? YearEnd { get; init; }

        /// <summary>
        /// Unique period readable text id, to be used in URLs.
        /// </summary>
        /// <example>late-baroque</example>
        public string Slug { get; init; } = "";

        /// <summary>
        /// Composers associated with the period.
        /// </summary>
        public Composer[] Composers { get; init; } = Array.Empty<Composer>();
    };
}

File: Site/Domain/Period.fs
/// Business logic for Period.
module Site.Domain.Period

open FSharp.Json
open Site.Domain.Composer
open Site.Postgres
open Site.Redis

/// Redis key for caching main page content
let redisKey = "opusclassical:periods"

/// Period when composer lived and worked, e.g. Late Baroque or Romanticism.
type Period =
    { id: int
      name: string
      yearStart: int
      yearEnd: int option
      slug: string // Unique period readable text id, to be used in URLs.
      composers: Composer list }

/// Returns all periods
let listPeriods () : Period list =
    match retrieveRedis redisKey with
    | Some c -> Json.deserialize<Period list> c
    | None ->
        let sql = "select json from periods_composers"
        let parameters = None

        let json =
            query (sql, parameters, jsonMapper) |> Async.RunSynchronously

        storeRedis (redisKey, json.Head, expire.Long) |> ignore

        json.Head |> Json.deserialize<Period list>
